File: src/main/java/clara/benchmark/interop/AccumInsert.java
package clara.benchmark.interop;

import clara.benchmark.Data;
import clara.benchmark.IBenchmark;
import clara.rules.QueryResult;
import clara.rules.RuleLoader;
import clara.rules.WorkingMemory;

import java.util.Collections;

public class AccumInsert implements IBenchmark<WorkingMemory> {

    @Override
    public String getName() {

        return "clara.accum-insert";
    }

    @Override
    public WorkingMemory getSession() throws Exception {
        return RuleLoader.loadRules("clara.benchmark.accum-insert");
    }

    @Override
    public Iterable<?> getFacts() throws Exception {
        return Data.getCustomers(100);
    }

    @Override
    public void run(WorkingMemory session, Iterable<Object> facts) throws Exception {

        Iterable<QueryResult> results = session.insert(facts)
                .fireRules()
                .query("clara.benchmark.accum-insert/get-discounts");

        // Iterate through results to ensure all work is done.
        for (QueryResult result: results)  {
            result.getResult("?percent");
            // System.out.println("percent: " + result.getResult("?percent"));
        }
    }
}

File: src/main/clojure/clara/benchmark/accum_insert.clj
(ns clara.benchmark.accum-insert
  (:require [clara.rules :refer :all]
            [clara.rules.accumulators :as acc])
  (:import [clara.benchmark.beans Discount Customer]))

(defrule vip-discount
   [?customers <- (acc/all) :from [Customer (= true vip)]]
   =>
   (apply insert! (for [customer ?customers]
                    (Discount. 10))))

(defquery get-discounts
  []
  [Discount (= ?percent percent)])